### PR TITLE
Feature: Add TrackSegment.getNextValidSegments() plugin API

### PR DESCRIPTION
**Pull request by markusklock**
Adds `TrackSegment.getNextValidSegments(rideId)` to the plugin API, allowing
plugins to query valid track piece transitions programmatically.

Previously, plugins that wanted to build rides programmatically had to hardcode
all track connection rules. This method exposes the game's internal track
compatibility logic to plugins.

## Changes
- Added `getNextValidSegments(rideId: number): TrackSegment[]` method
- Bumped plugin API version to 111
- Added changelog entry

Tested with a plugin that queries valid segments for various track pieces
on different ride types.

Solves:
https://github.com/OpenRCT2/OpenRCT2/discussions/25073

### CODE REVIEW

File: src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp [hunk 1]
```diff
@@ -298,4 +302,47 @@ std::string ScTrackSegment::getTrackPitchDirection() const
     return "flat";
 }
 
+std::vector<std::shared_ptr<ScTrackSegment>> ScTrackSegment::getNextValidSegments(int32_t rideId) const
```
Comment (ltsSmitty) [inline src/openrct2/scripting/bindings/ride/ScTrackSegment.cpp]: While this seems elegant and functional, why can't you use whatever actual implementation the game uses in the construction window when it's enabling/disabling the buttons?

I think there's a relatively low risk of them getting disconnected, but it'd make me happy to use the exact same function as the construction window. 

(For context I explored this exact same issue when I was trying to build a plugin to fully replace the ride construction window, and it was a challenge. So I have some thoughts on this.)

My implementation preference would be to refactor/extract the selection mechanism that the ride window uses and directly expose that to the API. Even if the plugin API is frozen for a time, you could do that refactor in a way that's ready to drop in once it is unfrozen.

**Comment by Basssiiie** [comment]
Hi, thank you for your submission. Unfortunately any changes to the plugin api are currently on hold, as we are working on replacing the underlying engine from Duktape to QuickJS, as can be seen in #23465.

Fortunately, that PR is in it's final phases, so hopefully it won't take too long anymore. Until then I will update the labels so we can keep track of it until the migration is merged. 🙂

**Comment by Basssiiie** [comment]
Hi @ZerxXxes 

You may or may not have noticed, but the QuickJS refactor has been merged, which means we are reopening plugin API contributions again. Would you still be interested in rebasing and adapting your PR to the new JS engine? Your contribution would be more than welcome. 

Thank you for your patience and looking forward to hearing from you. 🙂

**Comment by markusklock** [comment]
Hi @Basssiiie
I've rebased onto current develop and ported the implementation to the new quickjs-ng API. 
The C++ logic is unchanged and only the binding glue (signature, argument unpacking, result construction, registration) was rewritten to match the new patterns. 
Bumped kPluginApiVersion to 114 and moved the .d.ts entry to its new distribution/scripting/openrct2.d.ts location. 
Verified locally with a test-plugin that should catch most edge cases.
Ready for re-review.

File: distribution/changelog.txt [hunk 2]
```diff
@@ -2,6 +2,7 @@
 ------------------------------------------------------------------------
 - Feature: [#24242] [Plugin] Add ride-breakdown hooktype.
 - Feature: [#24879] [Plugin] Add methods for showing and hiding gridlines.
+- Feature: [Plugin] Add TrackSegment.getNextValidSegments(rideId) API to query valid track connections.
```
Comment (Basssiiie) [inline distribution/changelog.txt]: You can use the PR number to make CI happy about it, and backticks are usually used for literal function names.

```suggestion
- Feature: [#25840] [Plugin] Add `TrackSegment.getNextValidSegments(rideId)` API to query valid track connections.
```

I'll try a test later when I have some time. 🙂

**Comment by markusklock** [comment]
Made another commit to fix two issues in the matching logic:
1. Covered variants (flatCovered, etc.) were leaking into rides without RtdFlag::hasCoveredPieces. Now skipped via trackTypeIsCovered().
2. Inverted-variant rides (Flying, Lay-down, Multi-dimension) drew from a separate InvertedTrackPaintFunctions descriptor whose groups were ignored. The group lookup now checks all four applicable drawer entries (regular/covered × non-inverted/inverted), mirroring getCurrentTrackDrawerEntry in the construction code.

**Comment by Basssiiie** [comment]
Thanks for fixing those already! Here's some more things I noticed. 

For the train this straight-to-diagonal turn doesn't seem to have flat track or other diagonal turns as a next valid piece.
image

A vertical tower piece (not base) seems to have the tower base as a next valid piece.
image

The upwards quarter loop to inverted piece, and also the downward quarter loop, on the multi-dim gives no results.
image

And inverted diagonal banked turn, and various other diagonal elements.
image

Same for inverted inline twists on the multi-dim:
image

The compact inverted coaster shows turns on the vertical pieces (both up and down), which aren't possible, and it doesn't show any pieces that are actually valid (up90, down90, up90to60, down90to60 etc.)
image

The inverted impulse coaster doesn't show any next valid pieces on any of its vertical pieces.
image

Any element that has a flat unbanked exit on the Vertical Drop coaster also reports the Maze (id 101) as a possible next valid piece.
image

<details><summary>Click here for my test plugin.</summary>

```js
/// <reference path="../bin/openrct2.d.ts" />

function openResultWindow(trackSegment, rideId) {
    var rideName = map.getRide(rideId).name;
    var nextSegments = trackSegment.getNextValidSegments(rideId);

    var items = nextSegments.map(function (seg) {
        return [
            String(seg.type),
			TrackElemType[seg.type],
            seg.description,
            seg.turnDirection,
            seg.slopeDirection,
            String(seg.length)
        ];
    });

    ui.openWindow({
        classification: "test-next-valid-segments",
        title: "Next valid segments (" + nextSegments.length + ")",
        width: 600,
        height: 350,
        widgets: [
            {
                type: "label",
                x: 5,
                y: 20,
                width: 590,
                height: 14,
                text: "Picked: [" + trackSegment.type + "] " + TrackElemType[trackSegment.type] + ": " + trackSegment.description + "  —  Ride: " + rideName
            },
            {
                type: "listview",
                name: "segmentList",
                x: 5,
                y: 38,
                width: 590,
                height: 305,
                scrollbars: "vertical",
                isStriped: true,
                showColumnHeaders: true,
                columns: [
                    { header: "Type", width: 50 },
					{ header: "Name", width: 200 },
                    { header: "Description", width: 160 },
                    { header: "Curve", width: 60 },
                    { header: "Slope", width: 60 },
                    { header: "Length", width: 60 }
                ],
                items: items
            }
        ]
    });
}

function activatePickerTool() {
    ui.activateTool({
        id: "test-next-valid-segments-picker",
        cursor: "cross_hair",
        filter: ["ride"],
        onDown: function (e) {
            if (e.mapCoords && e.tileElementIndex !== undefined) {
                var tile = map.getTile(e.mapCoords.x / 32, e.mapCoords.y / 32);
                var element = tile.elements[e.tileElementIndex];
                if (element.type === "track") {
                    var trackType = element.trackType;
                    var rideId = element.ride;
                    var segment = context.getTrackSegment(trackType);
                    if (segment) {
                        openResultWindow(segment, rideId);
					}
                }
            }
        }
    });
}

registerPlugin({
    name: "test-next-valid-segments",
    version: "1.0.0",
    authors: "Basssiiie",
    type: "local",
    licence: "MIT",
    targetApiVersion: 77,
    main: function () {
        ui.registerMenuItem("Test next valid track segments", function () {
            activatePickerTool();
        });
    }
});

var TrackElemType =
[
	"Flat",
	"EndStation",
	"BeginStation",
	"MiddleStation",
	"Up25",
	"Up60",
	"FlatToUp25",
	"Up25ToUp60",
	"Up60ToUp25",
	"Up25ToFlat",
	"Down25",
	"Down60",
	"FlatToDown25",
	"Down25ToDown60",
	"Down60ToDown25",
	"Down25ToFlat",
	"LeftQuarterTurn5Tiles",
	"RightQuarterTurn5Tiles",
	"FlatToLeftBank",
	"FlatToRightBank",
	"LeftBankToFlat",
	"RightBankToFlat",
	"BankedLeftQuarterTurn5Tiles",
	"BankedRightQuarterTurn5Tiles",
	"LeftBankToUp25",
	"RightBankToUp25",
	"Up25ToLeftBank",
	"Up25ToRightBank",
	"LeftBankToDown25",
	"RightBankToDown25",
	"Down25ToLeftBank",
	"Down25ToRightBank",
	"LeftBank",
	"RightBank",
	"LeftQuarterTurn5TilesUp25",
	"RightQuarterTurn5TilesUp25",
	"LeftQuarterTurn5TilesDown25",
	"RightQuarterTurn5TilesDown25",
	"SBendLeft",
	"SBendRight",
	"LeftVerticalLoop",
	"RightVerticalLoop",
	"LeftQuarterTurn3Tiles",
	"RightQuarterTurn3Tiles",
	"LeftBankedQuarterTurn3Tiles",
	"RightBankedQuarterTurn3Tiles",
	"LeftQuarterTurn3TilesUp25",
	"RightQuarterTurn3TilesUp25",
	"LeftQuarterTurn3TilesDown25",
	"RightQuarterTurn3TilesDown25",
	"LeftQuarterTurn1Tile",
	"RightQuarterTurn1Tile",
	"LeftTwistDownToUp",
	"RightTwistDownToUp",
	"LeftTwistUpToDown",
	"RightTwistUpToDown",
	"HalfLoopUp",
	"HalfLoopDown",
	"LeftCorkscrewUp",
	"RightCorkscrewUp",
	"LeftCorkscrewDown",
	"RightCorkscrewDown",
	"FlatToUp60",
	"Up60ToFlat",
	"FlatToDown60",
	"Down60ToFlat",
	"TowerBase",
	"TowerSection",
	"FlatCovered",
	"Up25Covered",
	"Up60Covered",
	"FlatToUp25Covered",
	"Up25ToUp60Covered",
	"Up60ToUp25Covered",
	"Up25ToFlatCovered",
	"Down25Covered",
	"Down60Covered",
	"FlatToDown25Covered",
	"Down25ToDown60Covered",
	"Down60ToDown25Covered",
	"Down25ToFlatCovered",
	"LeftQuarterTurn5TilesCovered",
	"RightQuarterTurn5TilesCovered",
	"SBendLeftCovered",
	"SBendRightCovered",
	"LeftQuarterTurn3TilesCovered",
	"RightQuarterTurn3TilesCovered",
	"LeftHalfBankedHelixUpSmall",
	"RightHalfBankedHelixUpSmall",
	"LeftHalfBankedHelixDownSmall",
	"RightHalfBankedHelixDownSmall",
	"LeftHalfBankedHelixUpLarge",
	"RightHalfBankedHelixUpLarge",
	"LeftHalfBankedHelixDownLarge",
	"RightHalfBankedHelixDownLarge",
	"LeftQuarterTurn1TileUp60",
	"RightQuarterTurn1TileUp60",
	"LeftQuarterTurn1TileDown60",
	"RightQuarterTurn1TileDown60",
	"Brakes",
	"Booster",
	"Maze",
	"LeftQuarterBankedHelixLargeUp",
	"RightQuarterBankedHelixLargeUp",
	"LeftQuarterBankedHelixLargeDown",
	"RightQuarterBankedHelixLargeDown",
	"LeftQuarterHelixLargeUp",
	"RightQuarterHelixLargeUp",
	"LeftQuarterHelixLargeDown",
	"RightQuarterHelixLargeDown",
	"Up25LeftBanked",
	"Up25RightBanked",
	"Waterfall",
	"Rapids",
	"OnRidePhoto",
	"Down25LeftBanked",
	"Down25RightBanked",
	"Watersplash",
	"FlatToUp60LongBase",
	"Up60ToFlatLongBase",
	"Whirlpool",
	"Down60ToFlatLongBase",
	"FlatToDown60LongBase",
	"CableLiftHill",
	"ReverseFreefallSlope",
	"ReverseFreefallVertical",
	"Up90",
	"Down90",
	"Up60ToUp90",
	"Down90ToDown60",
	"Up90ToUp60",
	"Down60ToDown90",
	"BrakeForDrop",
	"LeftEighthToDiag",
	"RightEighthToDiag",
	"LeftEighthToOrthogonal",
	"RightEighthToOrthogonal",
	"LeftEighthBankToDiag",
	"RightEighthBankToDiag",
	"LeftEighthBankToOrthogonal",
	"RightEighthBankToOrthogonal",
	"DiagFlat",
	"DiagUp25",
	"DiagUp60",
	"DiagFlatToUp25",
	"DiagUp25ToUp60",
	"DiagUp60ToUp25",
	"DiagUp25ToFlat",
	"DiagDown25",
	"DiagDown60",
	"DiagFlatToDown25",
	"DiagDown25ToDown60",
	"DiagDown60ToDown25",
	"DiagDown25ToFlat",
	"DiagFlatToUp60",
	"DiagUp60ToFlat",
	"DiagFlatToDown60",
	"DiagDown60ToFlat",
	"DiagFlatToLeftBank",
	"DiagFlatToRightBank",
	"DiagLeftBankToFlat",
	"DiagRightBankToFlat",
	"DiagLeftBankToUp25",
	"DiagRightBankToUp25",
	"DiagUp25ToLeftBank",
	"DiagUp25ToRightBank",
	"DiagLeftBankToDown25",
	"DiagRightBankToDown25",
	"DiagDown25ToLeftBank",
	"DiagDown25ToRightBank",
	"DiagLeftBank",
	"DiagRightBank",
	"LogFlumeReverser",
	"SpinningTunnel",
	"LeftBarrelRollUpToDown",
	"RightBarrelRollUpToDown",
	"LeftBarrelRollDownToUp",
	"RightBarrelRollDownToUp",
	"LeftBankToLeftQuarterTurn3TilesUp25",
	"RightBankToRightQuarterTurn3TilesUp25",
	"LeftQuarterTurn3TilesDown25ToLeftBank",
	"RightQuarterTurn3TilesDown25ToRightBank",
	"PoweredLift",
	"LeftLargeHalfLoopUp",
	"RightLargeHalfLoopUp",
	"LeftLargeHalfLoopDown",
	"RightLargeHalfLoopDown",
	"LeftFlyerTwistUp",
	"RightFlyerTwistUp",
	"LeftFlyerTwistDown",
	"RightFlyerTwistDown",
	"FlyerHalfLoopUninvertedUp",
	"FlyerHalfLoopInvertedDown",
	"LeftFlyerCorkscrewUp",
	"RightFlyerCorkscrewUp",
	"LeftFlyerCorkscrewDown",
	"RightFlyerCorkscrewDown",
	"HeartLineTransferUp",
	"HeartLineTransferDown",
	"LeftHeartLineRoll",
	"RightHeartLineRoll",
	"MinigolfHoleA",
	"MinigolfHoleB",
	"MinigolfHoleC",
	"MinigolfHoleD",
	"MinigolfHoleE",
	"MultiDimInvertedFlatToDown90QuarterLoop",
	"Up90ToInvertedFlatQuarterLoop",
	"InvertedFlatToDown90QuarterLoop",
	"LeftCurvedLiftHill",
	"RightCurvedLiftHill",
	"LeftReverser",
	"RightReverser",
	"AirThrustTopCap",
	"AirThrustVerticalDown",
	"AirThrustVerticalDownToLevel",
	"BlockBrakes",
	"LeftBankedQuarterTurn3TileUp25",
	"RightBankedQuarterTurn3TileUp25",
	"LeftBankedQuarterTurn3TileDown25",
	"RightBankedQuarterTurn3TileDown25",
	"LeftBankedQuarterTurn5TileUp25",
	"RightBankedQuarterTurn5TileUp25",
	"LeftBankedQuarterTurn5TileDown25",
	"RightBankedQuarterTurn5TileDown25",
	"Up25ToLeftBankedUp25",
	"Up25ToRightBankedUp25",
	"LeftBankedUp25ToUp25",
	"RightBankedUp25ToUp25",
	"Down25ToLeftBankedDown25",
	"Down25ToRightBankedDown25",
	"LeftBankedDown25ToDown25",
	"RightBankedDown25ToDown25",
	"LeftBankedFlatToLeftBankedUp25",
	"RightBankedFlatToRightBankedUp25",
	"LeftBankedUp25ToLeftBankedFlat",
	"RightBankedUp25ToRightBankedFlat",
	"LeftBankedFlatToLeftBankedDown25",
	"RightBankedFlatToRightBankedDown25",
	"LeftBankedDown25ToLeftBankedFlat",
	"RightBankedDown25ToRightBankedFlat",
	"FlatToLeftBankedUp25",
	"FlatToRightBankedUp25",
	"LeftBankedUp25ToFlat",
	"RightBankedUp25ToFlat",
	"FlatToLeftBankedDown25",
	"FlatToRightBankedDown25",
	"LeftBankedDown25ToFlat",
	"RightBankedDown25ToFlat",
	"LeftQuarterTurn1TileUp90",
	"RightQuarterTurn1TileUp90",
	"LeftQuarterTurn1TileDown90",
	"RightQuarterTurn1TileDown90",
	"MultiDimUp90ToInvertedFlatQuarterLoop",
	"MultiDimFlatToDown90QuarterLoop",
	"MultiDimInvertedUp90ToFlatQuarterLoop",
	"RotationControlToggle",
	"FlatTrack1x4A",
	"FlatTrack2x2",
	"FlatTrack4x4",
	"FlatTrack2x4",
	"FlatTrack1x5",
	"FlatTrack1x1A",
	"FlatTrack1x4B",
	"FlatTrack1x1B",
	"FlatTrack1x4C",
	"FlatTrack3x3",
	"LeftLargeCorkscrewUp",
	"RightLargeCorkscrewUp",
	"LeftLargeCorkscrewDown",
	"RightLargeCorkscrewDown",
	"LeftMediumHalfLoopUp",
	"RightMediumHalfLoopUp",
	"LeftMediumHalfLoopDown",
	"RightMediumHalfLoopDown",
	"LeftZeroGRollUp",
	"RightZeroGRollUp",
	"LeftZeroGRollDown",
	"RightZeroGRollDown",
	"LeftLargeZeroGRollUp",
	"RightLargeZeroGRollUp",
	"LeftLargeZeroGRollDown",
	"RightLargeZeroGRollDown",
	"LeftFlyerLargeHalfLoopUninvertedUp",
	"RightFlyerLargeHalfLoopUninvertedUp",
	"LeftFlyerLargeHalfLoopInvertedDown",
	"RightFlyerLargeHalfLoopInvertedDown",
	"LeftFlyerLargeHalfLoopInvertedUp",
	"RightFlyerLargeHalfLoopInvertedUp",
	"LeftFlyerLargeHalfLoopUninvertedDown",
	"RightFlyerLargeHalfLoopUninvertedDown",
	"FlyerHalfLoopInvertedUp",
	"FlyerHalfLoopUninvertedDown",
	"LeftEighthToDiagUp25",
	"RightEighthToDiagUp25",
	"LeftEighthToDiagDown25",
	"RightEighthToDiagDown25",
	"LeftEighthToOrthogonalUp25",
	"RightEighthToOrthogonalUp25",
	"LeftEighthToOrthogonalDown25",
	"RightEighthToOrthogonalDown25",
	"DiagUp25ToLeftBankedUp25",
	"DiagUp25ToRightBankedUp25",
	"DiagLeftBankedUp25ToUp25",
	"DiagRightBankedUp25ToUp25",
	"DiagDown25ToLeftBankedDown25",
	"DiagDown25ToRightBankedDown25",
	"DiagLeftBankedDown25ToDown25",
	"DiagRightBankedDown25ToDown25",
	"DiagLeftBankedFlatToLeftBankedUp25",
	"DiagRightBankedFlatToRightBankedUp25",
	"DiagLeftBankedUp25ToLeftBankedFlat",
	"DiagRightBankedUp25ToRightBankedFlat",
	"DiagLeftBankedFlatToLeftBankedDown25",
	"DiagRightBankedFlatToRightBankedDown25",
	"DiagLeftBankedDown25ToLeftBankedFlat",
	"DiagRightBankedDown25ToRightBankedFlat",
	"DiagFlatToLeftBankedUp25",
	"DiagFlatToRightBankedUp25",
	"DiagLeftBankedUp25ToFlat",
	"DiagRightBankedUp25ToFlat",
	"DiagFlatToLeftBankedDown25",
	"DiagFlatToRightBankedDown25",
	"DiagLeftBankedDown25ToFlat",
	"DiagRightBankedDown25ToFlat",
	"DiagUp25LeftBanked",
	"DiagUp25RightBanked",
	"DiagDown25LeftBanked",
	"DiagDown25RightBanked",
	"LeftEighthBankToDiagUp25",
	"RightEighthBankToDiagUp25",
	"LeftEighthBankToDiagDown25",
	"RightEighthBankToDiagDown25",
	"LeftEighthBankToOrthogonalUp25",
	"RightEighthBankToOrthogonalUp25",
	"LeftEighthBankToOrthogonalDown25",
	"RightEighthBankToOrthogonalDown25",
	"DiagBrakes",
	"DiagBlockBrakes",
	"Down25Brakes",
	"DiagBooster",
	"DiagFlatToUp60LongBase",
	"DiagUp60ToFlatLongBase",
	"DiagFlatToDown60LongBase",
	"DiagDown60ToFlatLongBase",
	"LeftEighthDiveLoopUpToOrthogonal",
	"RightEighthDiveLoopUpToOrthogonal",
	"LeftEighthDiveLoopDownToDiag",
	"RightEighthDiveLoopDownToDiag",
	"DiagDown25Brakes",
	"Count",
]
```

</details>

**Comment by markusklock** [comment]
That's a neat test plugin, thanks!

I used it and started to investigate the bugs you found and I found that the previous approach duplicated only part of the construction-window logic, which made it easy for edge cases such as special pieces, covered variants, inverted variants, long-base rewrites, and direction-specific slope gates to diverge.
So this resulted in a larger commit than first anticipated.

To avoid two copies of kNextSelectedPiece and kSpecialElementsDropdownOrder, I moved both tables into a new core file TrackPieceSelection.
The new TrackPieceSelection helper centralises the selection pipeline and is used by both the UI-facing construction code and the scripting binding. 

I also added focused tests for the regressions/edge cases that motivated the restructure: special-piece overrides, vertical curve/slope gates, diagonal continuations, steep-slope direction splits, inverted variant roll handling, covered-piece substitution, and long-base rewrites.

**Comment by Basssiiie** [comment]
Thanks! It's kind of a hassle to test with so many possibilities, but managed to do one again. It looks like most are indeed fixed, but I still found a few issues.

The downward inverted quarter loop seems to still show nothing:
afbeelding

Left and right banked turns to diagonal on various coasters:
afbeelding
afbeelding

Upwards and downward sloped banked turns that end in a diagonal:
afbeelding
afbeelding

**Comment by markusklock** [comment]
Yeah, it really is not easy to go through and properly test all possible track pieces. Thanks for doing the work and finding those extra cases! 
The new failures came from two assumptions in the continuation filtering that were still too close to TED metadata rather than the construction UI state.
The inverted-variant case needed to normalize `rollEnd=upsideDown` the same way ride construction does after placing a piece, then use the variant scan only to choose the regular/inverted drawer for group gating.

For the banked diagonal cases, the main issue was that some TED group annotations do not line up with how the construction window actually enables those pieces.
I changed the checks to follow the construction categories instead (`flatRollBanking`, `slopeRollBanking`, `slopeCurveBanked`, `slopeCurveLargeBanked`) and added regression tests for the examples you found, plus a broader check.

**Comment by Basssiiie** [comment]
Hi, sorry for the delay. I have been away and busy these weeks. Unfortunately one of the first pieces I check is already lacking options. Is that correct?

image

The banked upwards diagonal turn should have more than 2 options for next pieces:

- The next diagonal turn piece (missing).
- Up, stay banked (present).
- Up, go unbanked (missing).
- Go flat, banked (missing).
- Go flat, unbanked (present).

Maybe you can double check the other ones one more time as well. If it's of any help, here's the save file I'm using:
[Six Flags old version with more track pieces.zip](https://github.com/user-attachments/files/28950272/Six.Flags.old.version.with.more.track.pieces.zip)

**Comment by markusklock** [comment]
Thanks for the save file, that will help to test it my self.
Trying to untangle this I found that even more stuff from the UI logic needs to be unified if this will make sense.
I will split the needed changes in to another PR first, before I revisit the needed scripting API changes.
Marking this as draft in the mean time.
